libsmtutil/CVC5Interface.cpp

```diff
@@ -0,0 +1,319 @@
+			auto nat = m_solver.mkTerm(cvc5::Kind::BITVECTOR_TO_NAT, {arguments[0]});
+			if (!intSort->isSigned)
+				return nat;
+
+			auto type = arguments[0].getSort();
+			smtAssert(type.isBitVector(), "");
+			auto size = type.getBitVectorSize();
+			// CVC5 treats all BVs as unsigned, so we need to manually apply 2's complement if needed.
+			auto extractOp = m_solver.mkOp(cvc5::Kind::BITVECTOR_EXTRACT, {size - 1, size - 1});
+			return m_solver.mkTerm(cvc5::Kind::ITE,
+				{m_solver.mkTerm(
+					cvc5::Kind::EQUAL,
+					{
+							m_solver.mkTerm(extractOp, {arguments[0]}),
+							m_solver.mkBitVector(1, uint64_t{0})
+					}
+				),
+				nat,
+				m_solver.mkTerm(
+					cvc5::Kind::NEG,
+					{m_solver.mkTerm(cvc5::Kind::BITVECTOR_TO_NAT, {m_solver.mkTerm(cvc5::Kind::BITVECTOR_NEG, {arguments[0]})})}
+				)}
+			);
+		}
+		else if (n == "select")
+			return m_solver.mkTerm(cvc5::Kind::SELECT, {arguments[0], arguments[1]});
+		else if (n == "store")
+			return m_solver.mkTerm(cvc5::Kind::STORE, {arguments[0], arguments[1], arguments[2]});
+		else if (n == "const_array")
+		{
+			shared_ptr<SortSort> sortSort = std::dynamic_pointer_cast<SortSort>(_expr.arguments[0].sort);
+			smtAssert(sortSort, "");
+			return m_solver.mkConstArray(CVC5Sort(*sortSort->inner), arguments[1]);
+		}
+		else if (n == "tuple_get")
+		{
+			shared_ptr<TupleSort> tupleSort = std::dynamic_pointer_cast<TupleSort>(_expr.arguments[0].sort);
+			smtAssert(tupleSort, "");
+			auto tt = m_solver.mkTupleSort(CVC5Sort(tupleSort->components));
+			cvc5::Datatype const& dt = tt.getDatatype();
+			size_t index = std::stoul(_expr.arguments[1].name);
+			cvc5::DatatypeSelector s = dt[0][index];
+			return m_solver.mkTerm(cvc5::Kind::APPLY_SELECTOR, {s.getTerm(), arguments[0]});
+		}
+		else if (n == "tuple_constructor")
+		{
+			shared_ptr<TupleSort> tupleSort = std::dynamic_pointer_cast<TupleSort>(_expr.sort);
+			smtAssert(tupleSort, "");
+			auto tt = m_solver.mkTupleSort(CVC5Sort(tupleSort->components));
+			cvc5::Datatype const& dt = tt.getDatatype();
+			cvc5::DatatypeConstructor c = dt[0];
+			return m_solver.mkTerm(cvc5::Kind::APPLY_CONSTRUCTOR, vector<cvc5::Term>{c.getTerm()} + arguments);
+		}
+
+		smtAssert(false);
+	}
+	catch (cvc5::CVC5ApiException const& _e)
+	{
+		smtAssert(false, _e.what());
+	}
+
+	smtAssert(false);
+
+	// FIXME: Workaround for spurious GCC 12.1 warning (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105794)
+	util::unreachable();
+}
+
+cvc5::Sort CVC5Interface::CVC5Sort(Sort const& _sort)
+{
+	switch (_sort.kind)
+	{
+	case Kind::Bool:
+		return m_solver.getBooleanSort();
+	case Kind::Int:
+		return m_solver.getIntegerSort();
+	case Kind::BitVector:
+		return m_solver.mkBitVectorSort(dynamic_cast<BitVectorSort const&>(_sort).size);
+	case Kind::Function:
+	{
+		FunctionSort const& fSort = dynamic_cast<FunctionSort const&>(_sort);
+		if (fSort.domain.empty()) // function sort in cvc5 must have nonempty domain
+			return CVC5Sort(*fSort.codomain);
```

Comment:
Interestingly, `cvc5` does not allow to create FunctionSort with empty domain (which was apparently allowed in `CVC4`. Therefore I decided to return just the codomain in this case, but this should be tested.

Reply:
Hmm, that's kind of funny. `Z3Interface::z3Sort` does not cover the case `Kind::Function` at all.
I will have to check how come `cvc` has to cover that and `Z3` does not.

Reply:
So, in `Z3Interface`, there is a special function `Z3Interface::declareFunction` which intercepts the call to `declareVariable` if the sort of the variable is `Function`.
This mechanism is missing from `CVC5Interface`.

But, I don't think this makes a difference. In the end, it makes a call to Z3's `context::function` giving it an empty domain.
And `cvc5` no longer allows an empty domain.